OpenRA.Mods.Common/Traits/World/ResourceRenderer.cs

```diff
@@ -95,13 +105,16 @@ void IWorldLoaded.WorldLoaded(World w, WorldRenderer wr)
 			}
 		}
 
-		protected void UpdateSpriteLayers(CPos cell, ISpriteSequence sequence, int frame, PaletteReference palette)
+		protected void UpdateSpriteLayers(CPos cell, Sprite sprite, PaletteReference palette, bool ignoreWorldTint)
 		{
 			foreach (var kv in spriteLayers)
 			{
+				if (palette != kv.Key)
+					continue;
```

Comment:
This doesn't look right. The `kv.Value.Clear(cell);` needs to run to make sure that the old sprite is cleared if the resource changes to something that uses a different palette.

Reply:
I can't clear the old sprite otherwise I see only the shadows.

Reply:
Let me be clearer: this will cause corrupted states where two (four if you count shadows) sprites will be rendered on the same cell.

Reply:
I think the resource layer only allows one resource per tile so this shouldn't be a problem?

Reply:
Like I said in my first comment, the problem occurs when the one resource in the tile changes from type A which uses palette X to type B using palette Y. Adding this check stops the `kv.Value.Clear(cell);` below from clearing sprite A, which means that both sprites A and B will be rendered even though the cell no longer contains A.

Reply:
I see that you have rewritten this block, but the new version still prevents the clear from being called...

Reply:
I have not found a solution for clear to clear too much:

![image](https://user-images.githubusercontent.com/756669/98447330-467bf480-2124-11eb-8eaa-ed4e10582c06.png)

Reply:
#18982 fixes the limitations in `TerrainSpriteLayer` that forced the awkward `ResourceRenderer` implementation.

Updating this PR to be based on the new plumbing should solve these kind of issues.

Reply:
That makes things worse:

![image](https://user-images.githubusercontent.com/756669/104088303-240e7080-5266-11eb-8783-83215f0c82e2.png)

I pushed it to https://github.com/Mailaender/OpenRA/tree/multi-sheet-resource-renderer as this pull request is in a working condition for quite some time now.

Reply:
The fundamental problem is that the order that you loop over a dictionary is undefined, so under the old code there is no way to control which layer is drawn on top of the other. This wasn't a problem when the trait guaranteed that different layers never rendered to the same cells, but is a problem now with shadows.

#18982 removed the limitations that forced the ResourceRenderer to use a dictionary in the first place, giving it back control of the rendering order. Your branch reverts all those fixes, so its no wonder it didn't work!

3d8d9a74ae7519a3d0fde57e9fa966d46aa536e1 shows what I meant, which is much simpler.

Reply:
The `ShadowPalette` is also a problem, since we're actively removing those from all traits as part of adding support for 32bit sprites. Adapting this PR to use tint-based shadows is easy enough, but requires changes to `TerrainSpriteLayer`. At this point, if you don't object, it would be easier if I take over and file a new PR for this feature.

Reply:
Actually, the Sequence-based shadow frames have never supported a separate shadow palette, so adding that here is not needed. Dropping the `ShadowPalette` definition simplifies this further, and avoids the need for any other changes: 8e20a15fce6f9c351ecd5ba8c50d11f4dab2a80e

Reply:
#18982 has been merged.